### ISSUE TITLE
`twisted.web.client.Agent` does not add a Host header if `None` (the default) is passed for the `headers` argument to `request`

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4131]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#4131|
|Type|defect|
|Created|2009-11-24 01:48:57Z|
|Branch|https://github.com/twisted/twisted/tree/host-header-4131|

`Agent.request` tries to make the _Host_ header optional.  It does this by adding it based on the request URL if it is not found in the headers object passed in.  However, if no headers object is passed in, it fails to perform that check.

<details><summary>Searchable metadata</summary>

```
trac-id__4131 4131
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_host_header_4131 branches/host-header-4131
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1259027337000000 1259027337000000
changetime__1259091901000000 1259091901000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [27596]) Branching to 'host-header-4131'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [27597]) Test the case where no headers are passed and fix the host handling in that case

refs [#4131](https://github.com/twisted/twisted/issues/4131)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|ivank commented|
|-|-|

1. In `test_hostProvided`, `body` is never used, so `body = object()` can be removed.

2. In `test_hostProvided`, local variable `request` is not used, so I suggest removing just the `request = `.

I've read it carefully, and everything else looks good.

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|ivank commented|
|-|-|

As an alternative fix for #2, `request =` could be changed to `result =` to be consistent with the other tests, which also do not use the variable.

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|ivank set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [27598]) Remove unused variables in new test method

refs [#4131](https://github.com/twisted/twisted/issues/4131)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set status to `closed`|
|-|-|

(In [27599]) Merge host-header-4131

Author: exarkun
Reviewer: ivank
Fixes: [#4131](https://github.com/twisted/twisted/issues/4131)

In `Agent.request`, apply the logic to insert a Host header even if no headers
were passed at all.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
